mesa_frames/abstract/space.py

```diff
@@ -733,3 +738,419 @@ def full_cells(self) -> DataFrame:
+            If 'moore', the neighborhood is the 8 cells around the center cell (up, down, left, right, and diagonals).
+            If 'von_neumann', the neighborhood is the 4 cells around the center cell (up, down, left, right).
+            If 'hexagonal', the neighborhood are 6 cells around the center cell distributed in a hexagonal shape.
+        """
+        super().__init__(model, capacity)
+        self._dimensions = dimensions
+        self._torus = torus
+        self._cells_col_names = [f"dim_{k}" for k in range(len(dimensions))]
+        self._center_col_names = [x + "_center" for x in self._cells_col_names]
+        self._agents = self._df_constructor(
+            columns=["agent_id"] + self._cells_col_names, index_col="agent_id"
+        )
+        self._cells = self._df_constructor(
+            columns=self._cells_col_names + ["capacity"],
+            index_cols=self._cells_col_names,
+        )
+        self._offsets = self._compute_offsets(neighborhood_type)
+        self._grid_capacity = self._generate_empty_grid(dimensions, capacity)
+        self._neighborhood_type = neighborhood_type
+
+    def get_directions(
+        self,
+        pos0: GridCoordinate | GridCoordinates | None = None,
+        pos1: GridCoordinate | GridCoordinates | None = None,
+        agents0: IdsLike | AgentContainer | Collection[AgentContainer] | None = None,
+        agents1: IdsLike | AgentContainer | Collection[AgentContainer] | None = None,
+        normalize: bool = False,
+    ) -> DataFrame:
+        result = self._calculate_differences(pos0, pos1, agents0, agents1)
+        if normalize:
+            result = result / self._df_norm(result)
+        return result
+
+    def get_distances(
+        self,
+        pos0: GridCoordinate | GridCoordinates | None = None,
+        pos1: GridCoordinate | GridCoordinates | None = None,
+        agents0: IdsLike | AgentContainer | Collection[AgentContainer] | None = None,
+        agents1: IdsLike | AgentContainer | Collection[AgentContainer] | None = None,
+    ) -> DataFrame:
+        result = self._calculate_differences(pos0, pos1, agents0, agents1)
+        return self._df_norm(result)
+
+    def get_neighbors(
+        self,
+        radius: int | Sequence[int],
+        pos: GridCoordinate | GridCoordinates | None = None,
+        agents: IdsLike | AgentContainer | Collection[AgentContainer] | None = None,
+        include_center: bool = False,
+    ) -> DataFrame:
+        if __debug__:
+            if pos is not None and agents is not None:
+                raise ValueError("Either pos or agents must be specified")
+        neighborhood_df = self.get_neighborhood(
+            radius=radius, pos=pos, agents=agents, include_center=include_center
+        )
+        return self._df_get_masked_df(
+            df=self._agents,
+            index_col="agent_id",
+            mask=neighborhood_df,
+            columns=self._agents.columns,
+        )
+
+    def get_cells(
```

Comment:
Come to think of it, it's strange when the method name is `get_cells` but the argument is `cells`.

Reply:
Yeah it's weird. 
I will change it to `mask` so it's clearer. I wanted to implement an indexable property for this and others properties (like neighborhood, neighbors ecc.) so that one could get the full DF by doing `space.cells` and a masked DF by doing `space.cells[coordinates]` but I don't think there's a way to do it without overriding the __getitem__ behavior of the DF so I think keeping a method is still the best way.

Reply:
I agree with the decision here for now. Ideally the DF is as vanilla as possible. Can't have an implementation that is too custom.